File: biddo_client/lib/core/controllers/ai.dart
import 'package:biddo/core/controllers/currencies.dart';
import 'package:easy_localization/easy_localization.dart';
import 'package:get/get.dart';

import '../repositories/ai.dart';
import 'account.dart';
import 'flash.dart';
import 'image_picker.dart';
import 'settings.dart';

class AiController extends GetxController {
  final settingsController = Get.find<SettingsController>();
  final imagePickerController = Get.find<ImagePickerController>();
  final aiRepository = Get.find<AIRepository>();
  final currenciesController = Get.find<CurrenciesController>();
  final accountController = Get.find<AccountController>();
  final flashController = Get.find<FlashController>();

  Future<GeneratedDataResult?> generateAuctionDataFromImages() async {
    var aiEnabled = settingsController.settings.value.aiEnabled;
    if (!aiEnabled) {
      flashController.showMessageFlash(tr('ai.ai_disabled'));
      return null;
    }

    var settings = settingsController.settings.value;
    var usedAiResponses = accountController.account.value.aiResponsesCount ?? 0;
    var maxAiResponsesReached =
        settings.maxAiResponsesPerUser <= usedAiResponses;

    if (maxAiResponsesReached) {
      flashController.showMessageFlash(tr('ai.max_ai_responses_reached'));
      return null;
    }

    var coinsToSpend = usedAiResponses < settings.freeAiResponses
        ? 0
        : settings.aiResponsesPriceInCoins;
    if (coinsToSpend > accountController.account.value.coins) {
      flashController.showMessageFlash(tr('ai.not_enough_coins'));
      return null;
    }

    var galleryAssets = await imagePickerController.getGalleryAssetPaths();
    var cameraAssets = imagePickerController.cameraAssets;

    var result = await aiRepository.generateData(
      currenciesController.selectedCurrency.value?.name['en'] ?? 'USD',
      galleryAssets,
      cameraAssets,
    );

    if (result == null) {
      return null;
    }

    accountController.account.value.aiResponsesCount = usedAiResponses + 1;
    accountController.account.value.coins -= coinsToSpend;
    accountController.account.refresh();

    return result;
  }
}

File: core/controllers/ai.ts
import { compressFileForAI } from '@/utils/compressor'
import { BiddoSettings } from '../domain/settings'
import { AiGeneratedDataResult, AiRepository } from '../repositories/ai'
import { AppStore } from '../store'

class AiController {
  public async generateAuctionDataFromImages(params: {
    appSettings: BiddoSettings
    assets: File[]
    currency: string
  }): Promise<AiGeneratedDataResult | null> {
    const { appSettings, assets, currency } = params
    const aiEnabled = appSettings.aiEnabled
    if (!aiEnabled) {
      throw new Error('AI_DISABLED')
    }

    const usedAiResponses = AppStore.accountData?.aiResponsesCount ?? 0
    const maxAiResponsesReached = appSettings.maxAiResponsesPerUser <= usedAiResponses
    if (maxAiResponsesReached) {
      throw new Error('MAX_AI_RESPONSES_REACHED')
    }

    const coinsToSpend =
      usedAiResponses < appSettings.freeAiResponses ? 0 : appSettings.aiResponsesPriceInCoins
    if (coinsToSpend > (AppStore.accountData?.coins ?? 0)) {
      throw new Error('NOT_ENOUGH_COINS')
    }

    const compressedAssets = await Promise.all(
      assets.map(async (asset) => {
        return compressFileForAI(asset)
      })
    )

    return await AiRepository.generateDescription({
      files: compressedAssets,
      currency,
    })
  }
}

const aiController = new AiController()
export { aiController as AiController }
